### ISSUE TITLE
Nodered keeps asking for password

### ISSUE THREAD

**Issue by Slayerrulez** [hassio-addons/addon-node-red#190]
Problem/Motivation
Can't login on the web interface

Expected behavior
fill out user/pass --> success

Actual behavior
Keeps asking for password, no error

Steps to reproduce
I do not know, just installing the new version?

Proposed changes
Please make it work again, 2.0.1 was fine.

Config;

{
  "credential_secret": "***",
  "dark_mode": false,
  "http_node": {
    "username": "***",
    "password": "***"
  },
  "http_static": {
    "username": "***",
    "password": "***"
  },
  "ssl": false,
  "certfile": "fullchain.pem",
  "keyfile": "privkey.pem",
  "require_ssl": false,
  "system_packages": [],
  "npm_packages": [],
  "init_commands": [],
  "log_level": "info"
}

**Comment by Slayerrulez** [comment]
What info do you need to solve this?

**Comment by MTrab** [comment]
Seeing the same behaviour when using panel_iframe with this config:

{
  "credential_secret": "MY_CREDENTIAL",
  "dark_mode": true,
  "http_node": {
    "username": "MY_USER",
    "password": "MY_PASS"
  },
  "http_static": {
    "username": "MY_USER",
    "password": "MY_PASS"
  },
  "ssl": false,
  "certfile": "fullchain.pem",
  "keyfile": "privkey.pem",
  "require_ssl": false,
  "system_packages": [],
  "npm_packages": [],
  "init_commands": [],
  "port": 1880,
  "log_level": "info"
}

**Comment by Slayerrulez** [comment]
@MTrab Good to hear I am not the only one with this problem.

**Comment by Slayerrulez** [comment]
Additional info;

My microsd was full because of the homeassistant.db , so I flashed the latest hassio on a new microsd card. And restored my backup from 0.89.2 which was the last backup I had.
Then I updated my hassio to the latest 0.9 anbd then I updated my nodered to 3.0.0
Everyhting went well, no errors, just only a one time connection via ingress, then nothing anymore, no matter which user/pass I filled out. Sometimes it takes 2 tries with a combination of HA user and nodered pass. But 99% it keeps spamming me for a user/pass.

**Comment by DelukaDZ** [comment]
Same problem here
Keeps asking for a login and password
Nothing in the log

**Comment by frenck** [comment]
I cannot reproduce this with all the attempts I've done.

Could somebody please start providing logs, add-on and system version details (as shown when the add-on starts), the browser and browser version used, the operating system your browser runs on and maybe even some details on your setup, like: Reverse proxies and configs?

These reports do not help me at this point.

**Comment by MTrab** [comment]
Reverse Proxy: NGINX
OS: Windows 10 1809
Browser: Chrome, IE, Edge, Firefox - all of them.

Log snippet from when trying to login from panel_iframe:
https://hastebin.com/kifewonahe.cs

Perhaps I can setup a test virtual Hass.IO with this error, any interest?

**Comment by DelukaDZ** [comment]
Add-on version: 3.0.0
 You are running the latest version of this add-on.
 System: HassOS 2.11  (armv7 / raspberrypi3)
 Home Assistant version: 0.91.3
 Supervisor version: 159

Setup node red
___________________________________________________________________________________________________________
{
  "credential_secret": "IkMisJeLuka",
  "dark_mode": false,
  "http_node": {
    "username": "xxxxxxx",
    "password": "xxxxxxx@xxxx"
  },
  "http_static": {
    "username": "xxxxxx",
    "password": "xxxxx@xxxxxxx"
  },
  "ssl": true,
  "certfile": "fullchain.pem",
  "keyfile": "privkey.pem",
  "require_ssl": true,
  "system_packages": [],
  "npm_packages": [],
  "init_commands": []
}
___________________________________________________________________________________________________________
I gave the problem in chrome version Versie 73.0.3683.10 and Firefox 66.0.2 (32-bits)
Running them on a win7 64bit machine with latest update
reinstalling and restart those not fix te problem
And this is the log after a restart
___________________________________________________________________________________________________________
[s6-init] making user provided files available at /var/run/s6/etc...exited 0.
[s6-init] ensuring user provided files have correct perms...exited 0.
[fix-attrs.d] applying ownership & permissions fixes...
[fix-attrs.d] done.
[cont-init.d] executing container initialization scripts...
[cont-init.d] 00-banner.sh: executing... 
-----------------------------------------------------------
 Hass.io Add-on: Node-RED
 Flow-based programming for the Internet of Things
-----------------------------------------------------------
 Add-on version: 3.0.0
 You are running the latest version of this add-on.
 System: HassOS 2.11  (armv7 / raspberrypi3)
 Home Assistant version: 0.91.3
 Supervisor version: 159
-----------------------------------------------------------
 Please, share the above information when looking for help
 or support in, e.g., GitHub, forums or the Discord chat.
-----------------------------------------------------------
[cont-init.d] 00-banner.sh: exited 0.
[cont-init.d] 01-log-level.sh: executing... 
[cont-init.d] 01-log-level.sh: exited 0.
[cont-init.d] nginx.sh: executing... 
[cont-init.d] nginx.sh: exited 0.
[cont-init.d] node-red.sh: executing... 
[19:45:13] INFO: Password is NOT in the Have I Been Pwned database! Nice!
[19:45:14] INFO: Password is NOT in the Have I Been Pwned database! Nice!
audited 20 packages in 3.38s
found 2 vulnerabilities (1 low, 1 moderate)
  run `npm audit fix` to fix them, or `npm audit` for details
[cont-init.d] node-red.sh: exited 0.
[cont-init.d] user.sh: executing... 
[cont-init.d] user.sh: exited 0.
[cont-init.d] done.
[services.d] starting services
[services.d] done.
[19:45:21] INFO: Starting Node-RED...
> hassio-addon-node-red@ start /opt
> node $NODE_OPTIONS node_modules/node-red/red.js "--settings" "/etc/node-red/config.js"
15 Apr 19:45:27 - [info] 
Welcome to Node-RED
===================
15 Apr 19:45:27 - [info] Node-RED version: v0.20.5
15 Apr 19:45:27 - [info] Node.js  version: v10.14.2
15 Apr 19:45:27 - [info] Linux 4.14.98-v7 arm LE
15 Apr 19:45:28 - [info] Loading palette nodes
15 Apr 19:45:35 - [info] Settings file  : /etc/node-red/config.js
15 Apr 19:45:35 - [info] Context store  : 'default' [module=memory]
15 Apr 19:45:35 - [info] User directory : /config/node-red/
15 Apr 19:45:35 - [warn] Projects disabled : editorTheme.projects.enabled=false
15 Apr 19:45:35 - [info] Flows file     : /config/node-red/flows.json
15 Apr 19:45:35 - [info] Server now running at http://127.0.0.1:46836/
[19:45:35] INFO: Starting NGinx...
15 Apr 19:45:35 - [info] Starting flows
15 Apr 19:45:36 - [info] Started flows
15 Apr 19:45:36 - [info] [mqtt-broker:Mosquitto ] Connected to broker: mqtt://localhost:1883
15 Apr 19:45:40 - [info] [server:Home Assistant] WebSocket Connected to http://hassio/homeassistant
___________________________________________________________________________________________________________

**Comment by frenck** [comment]
@MTrab Thanks, does your reverse proxy add additional authentication on its own?

**Comment by MTrab** [comment]
@frenck no - it passes everything as-is, only injecting a SSL certificate in front of every host.

**Comment by MTrab** [comment]
@frenck here you can see the host: https://hastebin.com/lezocesuli.nginx

**Comment by DelukaDZ** [comment]
Just done a simple test and after changing the port to 1881 (from 1880) i get into node red web interface
inside hassio i get a blank screen with the nodred banner on top

**Comment by DelukaDZ** [comment]
update 
after restarting chrome i have the webpage back
it now works inside chrome and firefox
from here it looks you cant use port 1880

**Comment by chewbaccadk** [comment]
> Just done a simple test and after changing the port to 1881 (from 1880) i get into node red web interface via firefox but via chrome the problem is still there

Just tried the same - changing post to 1881 from 1880 - and node red web interface is back - both in chrome and firefox. Thx @DelukaDZ

**Comment by frenck** [comment]
Wait what? wtf... 🤷‍♂️

**Comment by MTrab** [comment]
Cannot confirm - changed to port 1881 and problem persists.

**Comment by MTrab** [comment]
Oh wait - accessing the IP:1881 directly works. But not the proxied site (have changed the port in the proxy definition)

**Comment by frenck** [comment]
I've just set up a basic reverse proxy to my Node-RED using NGinx... and I hate to say, it just works.

My nginx config:

```
server {
  set $forward_scheme http;
  set $server "homeassistant";
  set $port 1880;

  listen 80;

  server_name nodered.demo.addons.community;
  access_log /proc/1/fd/1 proxy;

  location / {
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection "upgrade";
    proxy_http_version 1.1;
    add_header X-Served-By $host;
    proxy_set_header Host $host;
    proxy_set_header X-Forwarded-Scheme $scheme;
    proxy_set_header X-Forwarded-Proto  $scheme;
    proxy_set_header X-Forwarded-For $remote_addr;
    proxy_pass $forward_scheme://$server:$port;
  }
}
```

**Comment by MTrab** [comment]
Further tests get it down to something about the certificate injection part.
Again - this worked until upgrade to version 3.0.

**Comment by frenck** [comment]
Well, one big difference @MTrab, from the Nginx config you've shared, you are not passing the scheme and proto headers... which is kinda essential for an application running behind a reverse proxy.

Yes, the add-on updated, changed and gained new features. But I feel to see issues with the questions asked. Your configuration for a reverse proxy contains some weird things IMHO. Given the above NGinx config, simple, basic requirements, just works.

I'm not able to break it in any way (well there are things broken, but we are talking about the login dialog).

Meanwhile, have you checked the Hass.io system logs for hints? Anything unusual there?

**Comment by MTrab** [comment]
Nothing unusual in any logs.

Not sure what wierd things you are thinking of.
The only difference from your example is that I inject a SSL cert and redirects all http to https.

**Comment by frenck** [comment]
@MTrab Nope, you are missing all kinds of proxy headers, which the remote application uses to determine its output.

**Comment by MTrab** [comment]
@frenck added the 2 headers I see you have, that I don't - still no joy.

**Comment by Rink76** [comment]
I experienced the same after upgrading to 3.0.0 port 1880 wasn't active anymore. Initially added a port mapping 80 <> 1880 within the network section. This got me past the first authentication box, but after that NodeRed kept asking for username/password. After changing the port to 1881 I'm able to access and authenticate successfully (https://hassioip:1881)

**Comment by MTrab** [comment]
I think I give up.
Might just have to get used to Hass.IO doubleing up on the interface, until Hass.IO add-ons have a better (old) way of integrate with panel_iframe :(
![Udklip](https://user-images.githubusercontent.com/14296463/56159493-18c0df00-5fc5-11e9-8efa-126d7d1b548e.JPG)

**Comment by frenck** [comment]
@Rink76 Thanks for letting me know.

@MTrab That is a totally different issue. I would really appreciate if we could stay on the topic inside GitHub issues.

**Comment by MTrab** [comment]
@frenck the issue I see here is a way to get round this  other issue.

**Comment by MTrab** [comment]
Working for me now. Disabled all access to node-red except ingress and used panel_custom to integrate a link in the left hand panel.
Not very well documented, but that's a whole other story.

**Comment by frenck** [comment]
For people who opened direct access by enabling and setting the port to 1880 and experiencing repeated login cycles. This might be a solution to resolve it:

![nodered-login-issues](https://user-images.githubusercontent.com/195327/56173198-ff338d80-5fec-11e9-9215-cf2d55e53704.gif)

It seems like node-red stored a piece of information about the login system used in versions of the node-red add-on before v3.0.0 (e.g., v2.0.2).

This authentication method does not exist anymore but nevertheless, throws off Node-RED so it seems.
Removing this manually helps in all the tests I’ve done tonight.

**Comment by nicr4wks** [comment]
@frenck Thankyou can confirm this worked for me using port 1880.

**Comment by CyaOnDaNet** [comment]
I can also confirm this fixed my login on port 1880

**Comment by Rink76** [comment]
@frenck Thnx! Indeed, this solves the issue.  I'm able to access NodeRed on port 1880 again

**Comment by rafuz** [comment]
There is still an issue about auth that is not fixed by this workaround.
@frenck has found the problem, will be patched in a new release

**Comment by Slayerrulez** [comment]
@frenck Really nice troubleshooting skills! Thank you for your time to find the issue and everybody for your quick responses.

**Comment by Slayerrulez** [comment]
@frenck Now the nodered website is showing up but the dashboard ui isn't. Related or new bug report?

**Comment by frenck** [comment]
@Slayerrulez New bug report.

**Comment by sretalla** [comment]
I had thought there was a related problem to this with the add-on failing to load, but it was me deciding to try dark_mode (which causes the add-on to not load, by the way).

I am finding in this case that the continual prompt is solved by using the old static credentials I had in the past, not the ones from Home Assistant that have been working prior to the upgrade.

I note that I still have the following in settings.js:
```    // property can be used. See http://nodered.org/docs/security.html for details.
    adminAuth: {
      type: "credentials",
      users: [
                {
                  username: "admin",
                  password: "encoded string here",
                  permissions: "*"
                }
      ]
    },

    // To password protect the node-defined HTTP endpoints (httpNodeRoot), or
    // the static content (httpStatic), the following properties can be used.
    // The pass field is a bcrypt hash of the password.
    // See http://nodered.org/docs/security.html#generating-the-password-hash
    httpNodeAuth: {user:"admin",pass:"encoded string here."},```

**Comment by frenck** [comment]
@sretalla Correct, the upcoming release of the add-on will handle those setting automatically.

**Comment by frenck** [comment]
Released Node-RED v3.0.1, which addresses the final piece of this login issue.

If you are still having login issue, please look at the resolution in this comment:

https://github.com/hassio-addons/addon-node-red/issues/190#issuecomment-483463573

Thank you all for being so helpful, and above all, sorry for causing this issue.

**Comment by ArnoutVerbeken** [comment]
You rule.  Thanks for everything you do in the hass community.  You give the steriods to our hass with your shop full of addons.
